src/modules/interface/controller_indi.h

```diff
@@ -0,0 +1,125 @@
+/**
+ *    ||          ____  _ __
+ * +------+      / __ )(_) /_______________ _____  ___
+ * | 0xBC |     / __  / / __/ ___/ ___/ __ `/_  / / _ \
+ * +------+    / /_/ / / /_/ /__/ /  / /_/ / / /_/  __/
+ *  ||  ||    /_____/_/\__/\___/_/   \__,_/ /___/\___/
+ *
+ * Crazyflie control firmware
+ *
+ * Copyright (C) 2011-2016 Bitcraze AB
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, in version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * controller_indi.h - INDI Controller Interface
+ */
+#ifndef __CONTROLLER_INDI_H__
+#define __CONTROLLER_INDI_H__
+
+#include "stabilizer_types.h"
+#include "filter.h"
+#include "math3d.h"
+
+#define PI 3.14159265358979323846f
+
+//The G values are scaled to avoid numerical problems during the estimation
+#define INDI_EST_SCALE 0.001f
+
+// these parameters are used in the filtering of the angular acceleration
+#define STABILIZATION_INDI_FILT_CUTOFF 8.0f
+
+// the yaw sometimes requires more filtering
+#define STABILIZATION_INDI_FILT_CUTOFF_R STABILIZATION_INDI_FILT_CUTOFF
+
+#define STABILIZATION_INDI_ESTIMATION_FILT_CUTOFF 4.0f
+
+#define PERIODIC_FREQUENCY 512.0f
+
+// these parameters are used in the filtering of the angular acceleration
+#define STABILIZATION_INDI_MAX_RATE 6.0f
+#define STABILIZATION_INDI_MAX_R 120.0f
+#define STABILIZATION_INDI_G1_P 0.0639f
+#define STABILIZATION_INDI_G1_Q 0.0361f
+#define STABILIZATION_INDI_G1_R 0.0022f
+#define STABILIZATION_INDI_G2_R 0.1450f
+#define STABILIZATION_INDI_REF_ERR_P 600.0f
+#define STABILIZATION_INDI_REF_ERR_Q 600.0f
+#define STABILIZATION_INDI_REF_ERR_R 600.0f
+#define STABILIZATION_INDI_REF_RATE_P 28.0f
+#define STABILIZATION_INDI_REF_RATE_Q 28.0f
+#define STABILIZATION_INDI_REF_RATE_R 28.0f
+#define STABILIZATION_INDI_USE_ADAPTIVE false
+#define STABILIZATION_INDI_ADAPTIVE_MU 0.0001f
+#define STABILIZATION_INDI_FULL_AUTHORITY false
+#define STABILIZATION_INDI_ACT_DYN_P 0.1f
+#define STABILIZATION_INDI_ACT_DYN_Q 0.1f
+#define STABILIZATION_INDI_ACT_DYN_R 0.1f
```

Comment:
This we should determine with flight data, or better with an actuator identification setup. It is the discrete coefficient, so it actually depends on the control frequency. I think these actuators will have a (continuous) first order time constant of around 1/16, you can find the discrete value with a zero order hold with the correct control frequency.

Reply:
Did you mean we will calculate these constants with flight data, with the help of the SD-Card, or that we can calculate them with the control frequency? I got confused here, sorry.

Reply:
I mean we should determine these dynamics, but until we do, we can assume (based on experience) A(s) = 1/(tau*s+ 1). To get the discrete version of this with a zero order hold, you have to know ATTITUDE_RATE. For instance with Matlab you can do c2d(A, 1/ATTITUDE_RATE).

Reply:
>>> A = control.TransferFunction([1],[1/16,1])
>>> 
>>> 
>>> Ad = control.matlab.c2d(A,1/500)
>>> Ad

  0.03149
----------
z - 0.9685

dt = 0.002

Is this what you meant?

Reply:
Yes!

Reply:
Nice.

However, I still do not understand pragmatically what we can do with this discrete version in hands, or if/how it may help us determine the dynamics parameter.

Reply:
`#define STABILIZATION_INDI_ACT_DYN_R 0.03149f` 😉

Reply:
Alright, got it! :+1: